src/transpose_z_to_y.f90

```diff
@@ -476,4 +429,182 @@ subroutine mem_merge_zy_complex(in, n1, n2, n3, out, iproc, dist, decomp)
 
    end subroutine mem_merge_zy_complex
 
+!!! Utility subroutines
+    
+   subroutine rearrange_sendbuf_real(src, wk1, decomp)
+
+      real(mytype), dimension(:, :, :), intent(IN) :: src
+      TYPE(DECOMP_INFO), intent(IN) :: decomp
+      real(mytype), dimension(:), intent(out) :: wk1
+#if defined(_GPU)
+      attributes(device) :: wk1
+
+      integer :: istat
+#endif
+
+      integer :: s1, s2, s3
+
+      s1 = SIZE(src, 1)
+      s2 = SIZE(src, 2)
+      s3 = SIZE(src, 3)
+
+#ifdef EVEN
+      if (.not. decomp%even) then
+         call mem_split_zy_real(src, s1, s2, s3, wk1, dims(2), &
+                                decomp%z2dist, decomp)
+      end if
+#else
+      ! note the src array is suitable to be a send buffer
+      ! so no split operation needed
+
+#if defined(_GPU)
+      !$acc host_data use_device(src)
+      istat = cudaMemcpy(wk1, src, s1 * s2 * s3, cudaMemcpyDeviceToDevice)
+      !$acc end host_data
+      if (istat /= 0) call decomp_2d_abort(__FILE__, __LINE__, istat, "cudaMemcpy2D")
+#else
+      associate(foo => wk1, bar => decomp)
+      end associate
+#endif
+
+#endif
+   end subroutine rearrange_sendbuf_real
+
+   subroutine define_recvbuf_real(src, decomp, wk1, wk2)
```

Comment:
We could add some comments but otherwise the subroutine looks ok. See comment above about the name of the subroutine.